asdf/asdf.py

```diff
@@ -616,14 +617,23 @@ def _open_asdf(cls, self, fd, uri=None, mode='r',
                    _force_raw_types=False,
                    strict_extension_check=False,
                    ignore_missing_extensions=False,
-                   validate_on_read=True):
+                   validate_on_read=NotSet):
```

Comment:
Since `None` is falsy, I thought it best to make the new default something unambiguous.

Reply:
I see.  You're using it to detect if the optional argument has been passed to the function.

Since default values get filled in python when the function is defined, maybe the [following design pattern](https://stackoverflow.com/questions/14749328/how-to-check-whether-optional-function-parameter-is-set) works?

```
def configure(converter_providers=None, validate_on_read=None): 
    if converter_providers is configure.__defaults__[0]: 
        print("cp default") 
    else: 
        print("cp passed in call") 
    if validate_on_read is configure.__defaults__[1]: 
        print("vor default") 
    else: 
        print("vor passed in call") 
```

This gives a slightly more readable API for the user (at the expense of weird code in the function itself to handle it).

Reply:
In python, generally optional arguments that don't do anything unless something is passed to them are given `None`.  I think this is much more common than some custom type of which I don't (as a user) know what it means.  In the context of function args, `None` is not falsy in python.  It merely means, "does not apply".

Reply:
I just mean, we use the value of the flag like this:

```python
if validate_on_read:
  ...
```
so if someone had been passing `None` for some reason, it will have the opposite effect that it did in the previous version of asdf if we start treating `None` as "use the default value of `True`".  Too fine a point to worry about?

Reply:
Nobody should be passing `validate_on_read=None`, as it was previously a boolean in the API.  If they were passing `None` for some reason, it was not explicitly supported.

Reply:
Thinking about this a little more, what we actually want to detect is the fact that someone is calling this method with the `validate_on_read` keyword.  When they do that, we emit a deprecation warning to save them from an error later when the keyword disappears.

So maybe the best option here is to use `**kwargs`, check for presence of "validate_on_read" in that dict, and use it but warn appropriately.  In this case we don't care that the method signature will be missing `validate_on_read`, that's actually a virtue because we're removing the keyword and it will discourage people from using it.